### PR TITLE
ENH: `sklearnex` verbose mode messenger

### LINKED ISSUE #1416: Cannot enable GPU implementation for `NearestNeighbors(algorithm="brute")`
**To Reproduce**

I copied the snippet from the documentation for DBSCAN and it seems to work as expected:

```python
>>> import numpy as np
>>> from sklearnex import patch_sklearn, config_context
... patch_sklearn()
... 
... from sklearn.cluster import DBSCAN
... 
... X = np.array([[1., 2.], [2., 2.], [2., 3.],
...             [8., 7.], [8., 8.], [25., 80.]], dtype=np.float32)
... with config_context(target_offload="gpu:0"):
...    clustering = DBSCAN(eps=3, min_samples=2).fit(X)
... 
Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)
INFO:sklearnex: sklearn.cluster.DBSCAN.fit: running accelerated version on GPU
```

but then, swapping `DBSCAN` for `NearestNeighbors` does not seem to work (it still uses the CPU despite the `config_context`:

```python
>>> from sklearn.neighbors import NearestNeighbors
>>> with config_context(target_offload="gpu:0"):
...     NearestNeighbors(algorithm="brute", n_neighbors=2).fit(X).kneighbors(X)
... 
INFO:sklearnex: sklearn.neighbors.NearestNeighbors.fit: running accelerated version on CPU
INFO:sklearnex: sklearn.neighbors.NearestNeighbors.kneighbors: running accelerated version on CPU
```

**Expected behavior**

I would have expected the message to show that it can dispatch on GPU because oneDAL seem to have specialized GPU implementation for this algorithm:

https://github.com/oneapi-src/oneDAL/pull/705

**Environment:**
 - OS: Linux idc-beta-batch-pvc-node-04 5.15.0-79-generic #86-Ubuntu SMP Mon Jul 10 16:07:21 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux (Intel devcloud)
 - Packages installed from the `intel` conda channel.

There is a GPU reachable by Python on this machine:

```python
>>> import dpctl                                                                                                                                                                                                  
>>> dpctl.get_devices()                                                                                                                                                                                           
[<dpctl.SyclDevice [backend_type.opencl, device_type.cpu,  Intel(R) Xeon(R) Platinum 8480+] at 0x152b985cbe30>, <dpctl.SyclDevice [backend_type.opencl, device_type.accelerator,  Intel(R) FPGA Emulation Device] 
at 0x152b985cb2b0>, <dpctl.SyclDevice [backend_type.level_zero, device_type.gpu,  Intel(R) Data Center GPU Max 1100] at 0x152b985c8cf0>]
```

### PR DESCRIPTION
**Pull request by samir-nasibli**
# Description
`sklearnex` verbose mode messenger update:
* sklearnex own `PatchingConditionsChain` inherited from daal4py's one
* updated `_get_backend` and `dispatch` functions from `_device_offload` module for the new logic of patching.
* Added usage of `PatchingConditionsChain` into preview and non-preview sklearnex algos, with onedal4py backend
* fixes #1416 

# TODO:
* adding additional tests for this messenger. Could be done in separate PR

### CODE REVIEW

File: sklearnex/_device_offload.py [hunk 1]
```diff
@@ -147,22 +147,31 @@ def _get_backend(obj, queue, method_name, *data):
...
+        if patching_status.get_status():
+            return "onedal", queue, patching_status
+        else:
+            if allow_fallback_to_host:
+                patching_status = obj._onedal_cpu_supported(method_name, *data)
+                if patching_status.get_status():
+                    return "onedal", None, patching_status
+                else:
+                    return "sklearn", None, patching_status
+            else:
```
Comment (samir-nasibli) [inline sklearnex/_device_offload.py:173]: Logic will be carefully updated.

File: sklearnex/_utils.py [hunk 2]
```diff
@@ -20,9 +20,35 @@
...
+from daal4py.sklearn._utils import (
+    PatchingConditionsChain as daal4py_PatchingConditionsChain,
+)
 from daal4py.sklearn._utils import daal_check_version
 
 
+class PatchingConditionsChain(daal4py_PatchingConditionsChain):
+    def get_status(self, logs=False, queue=None):
+        if logs:
+            self.write_log(queue)
```
Comment (samir-nasibli) [inline sklearnex/_utils.py:32]: will be removed.

**Review by KulikovNikita — COMMENTED** [review]
In general looks fine to me

File: sklearnex/preview/ensemble/extra_trees.py [hunk 3]
```diff
@@ -593,26 +594,22 @@ def _onedal_gpu_supported(self, method_name, *data):
...
+                        not sp.issparse(X),
+                        "X is sparse. Sparse input is not supported.",
+                    ),
+                    (self.warm_start is False, "Warm start is not supported."),
+                    (
+                        daal_check_version((2023, "P", 100)),
+                        "ExtraTrees supported starting from oneDAL version 2023.1",
+                    ),
+                ]
+            )
```
Comment (samir-nasibli) [inline sklearnex/preview/ensemble/extra_trees.py:610]: @icfaust please check it

File: sklearnex/_utils.py [hunk 4]
```diff
@@ -20,9 +20,33 @@
...
 import warnings
 
+from daal4py.sklearn._utils import (
+    PatchingConditionsChain as daal4py_PatchingConditionsChain,
+)
 from daal4py.sklearn._utils import daal_check_version
 
 
+class PatchingConditionsChain(daal4py_PatchingConditionsChain):
+    def get_status(self, queue=None):
```
Comment (Alexsandruss) [inline sklearnex/_utils.py:30]: `queue` is unused.
```suggestion
    def get_status(self):
```

File: sklearnex/neighbors/lof.py [hunk 5]
```diff
@@ -236,10 +237,18 @@ def fit_predict(self, X, y=None):
             )
 
         def _onedal_gpu_supported(self, method_name, *data):
-            return True
+            class_name = self.__class__.__name__
+            patching_status = PatchingConditionsChain(
+                f"sklearn.cluster.{class_name}.{method_name}"
```
Comment (Alexsandruss) [inline sklearnex/neighbors/lof.py:242]:
```suggestion
                f"sklearn.neighbors.{class_name}.{method_name}"
```

File: sklearnex/neighbors/lof.py [hunk 6]
```diff
@@ -236,10 +237,18 @@ def fit_predict(self, X, y=None):
...
+            patching_status = PatchingConditionsChain(
+                f"sklearn.cluster.{class_name}.{method_name}"
+            )
+            return patching_status
 
         def _onedal_cpu_supported(self, method_name, *data):
-            return True
+            class_name = self.__class__.__name__
+            patching_status = PatchingConditionsChain(
+                f"sklearn.cluster.{class_name}.{method_name}"
```
Comment (Alexsandruss) [inline sklearnex/neighbors/lof.py:249]:
```suggestion
                f"sklearn.neighbors.{class_name}.{method_name}"
```

**Review by Alexsandruss — CHANGES_REQUESTED** [review]

File: sklearnex/_utils.py [hunk 7]
```diff
@@ -20,9 +20,33 @@
...
+        return self.patching_is_enabled
+
+    def write_log(self, queue=None):
```
[the remark this answers is not in the document]
Comment (samir-nasibli) [reply sklearnex/_utils.py:33]: I think we should have different implementations. At least we have different flows there

File: sklearnex/neighbors/lof.py (hunk 5 above)
Comment (samir-nasibli) [reply sklearnex/neighbors/lof.py:242]: Goof catch. thanks

File: sklearnex/neighbors/lof.py (hunk 6 above)
Comment (samir-nasibli) [reply sklearnex/neighbors/lof.py]: fixed.

File: sklearnex/_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Alexsandruss) [reply sklearnex/_utils.py:33]: No, branches of `get_patch_message` functions can be easily stacked and it's the only difference in implementations

**Review by Alexsandruss — DISMISSED** [review]
Merge `get_patch_message` functions:
```python
def get_patch_message(s, queue=None):
    if s == "daal":
        message = "running accelerated version on "
        if oneapi_is_available:
            dev = _get_device_name_sycl_ctxt()
            if dev == "cpu" or dev is None:
                message += "CPU"
            elif dev == "gpu":
                message += "GPU"
            else:
                raise ValueError(
                    f"Unexpected device name {dev}." " Supported types are cpu and gpu"
                )
        else:
            message += "CPU"
    elif s == "onedal":
        message = "running accelerated version on "
        if queue is not None:
            if queue.sycl_device.is_gpu:
                message += "GPU"
            elif queue.sycl_device.is_cpu:
                message += "CPU"
            else:
                raise RuntimeError("Unsupported device")
        else:
            message += "CPU"
    elif s == "sklearn":
        message = "fallback to original Scikit-learn"
    elif s in ["sklearn_after_daal", "sklearn_after_daal"]:
        message = "failed to run accelerated version, fallback to original Scikit-learn"
    else:
        raise ValueError(
            f"Invalid input - expected one of 'daal', 'onedal', 'sklearn',"
            f" 'sklearn_after_daal', 'sklearn_after_onedal', got {s}"
        )
    return message
```

**Comment by samir-nasibli** [comment]
> Merge `get_patch_message` functions:
> 
> ```python
> def get_patch_message(s, queue=None):
>     if s == "daal":
>         message = "running accelerated version on "
>         if oneapi_is_available:
>             dev = _get_device_name_sycl_ctxt()
>             if dev == "cpu" or dev is None:
>                 message += "CPU"
>             elif dev == "gpu":
>                 message += "GPU"
>             else:
>                 raise ValueError(
>                     f"Unexpected device name {dev}." " Supported types are cpu and gpu"
>                 )
>         else:
>             message += "CPU"
>     elif s == "onedal":
>         message = "running accelerated version on "
>         if queue is not None:
>             if queue.sycl_device.is_gpu:
>                 message += "GPU"
>             elif queue.sycl_device.is_cpu:
>                 message += "CPU"
>             else:
>                 raise RuntimeError("Unsupported device")
>         else:
>             message += "CPU"
>     elif s == "sklearn":
>         message = "fallback to original Scikit-learn"
>     elif s in ["sklearn_after_daal", "sklearn_after_daal"]:
>         message = "failed to run accelerated version, fallback to original Scikit-learn"
>     else:
>         raise ValueError(
>             f"Invalid input - expected one of 'daal', 'onedal', 'sklearn',"
>             f" 'sklearn_after_daal', 'sklearn_after_onedal', got {s}"
>         )
>     return message
> ```

That is bad practice. Daal4py shouldn\t know about sklearnex impl specifics.
Also skleearnex falls to the daal4py, that uses it is own logger.

**Comment by Alexsandruss** [comment]
> > Merge `get_patch_message` functions:
> > ```python
> > def get_patch_message(s, queue=None):
> >     ...
> > ```
> 
> That is bad practice. Daal4py shouldn\t know about sklearnex impl specifics. Also skleearnex falls to the daal4py, that uses it is own logger.

Logger is common functionality of daal4py and sklearnex. Thus, it should be aware of both modules. The only problem it's stored in one of them instead of separate place

**Comment by samir-nasibli** [comment]
> Logger is common functionality of daal4py and sklearnex. Thus, it should be aware of both modules. The only problem it's stored in one of them instead of separate place

Sure. Thank you for your thoughts.
We use different instances anyway and they are related, only the behavior of some methods is changed for onedal4py backed specifics.
daal4py doesn't need to know about the `sklearnex` implementation specifics.

**Comment by Alexsandruss** [comment]
> > Logger is common functionality of daal4py and sklearnex. Thus, it should be aware of both modules. The only problem it's stored in one of them instead of separate place
> 
> Sure. Thank you for your thoughts. We use different instances anyway and they are related, only the behavior of some methods is changed for onedal4py backed specifics. daal4py doesn't need to know about the `sklearnex` implementation specifics.

We used *one instance* before this PR and there was no need for another. `daal4py` was not aware of `sklearnex` specifics, common logger was just stored in `daal4py` namespace.

**Comment by samir-nasibli** [comment]
> > > Logger is common functionality of daal4py and sklearnex. Thus, it should be aware of both modules. The only problem it's stored in one of them instead of separate place
> > 
> > 
> > Sure. Thank you for your thoughts. We use different instances anyway and they are related, only the behavior of some methods is changed for onedal4py backed specifics. daal4py doesn't need to know about the `sklearnex` implementation specifics.
> 
> We used _one instance_ before this PR and there was no need for another. `daal4py` was not aware of `sklearnex` specifics, common logger was just stored in `daal4py` namespace.

I see what do you mean. So we used to have it, but it causes several problems. That is why this impl just inherited daal4py one functionality and just updates two methods, that should be stored on sklearnex side, since it is specific.

**Pull request merged**